Documentation (#9)
DevOps-MBSE/aac-spec: docs/source/project_documentation/specifications.md

# Specifications Plugin

> The `Specifications` plugin traces `Requirement`s of provided `Specification`s.  It does this through the `spec-csv` command, which generates a .csv file for a `Specification` which tracks its `Requirement`s.

Reviewer: So, remember the code highlighting of `using the back ticks` is only necessary if you are referencing code components. 

```suggestion
The Specifications Plugin traces requirements of provided modeled specifications.  It does this through the `spec-csv` command, which generates a *.csv file for a `req-spec` which traces its `requirements`.
```

## Spec-Csv Command
```bash
aac spec-csv architecture-file.aac output_directory
```

### Arguments

#### Architecture File

> The AaC file containing the `Specification`s and `Requirement`s definitions.

Reviewer: ```suggestion
The AaC file containing the `req-spec` and `req` definitions for a system, sub-system, or component.
```

#### Output Directory

> The directory in which the .csv file will be generated.

Reviewer: ```suggestion
The directory in which the *.csv file(s) will be generated.
```

## Plugin Usage Example

An Architecture File, `architecture_file.aac`, contains the following definitions:
```{eval-rst}
.. literalinclude:: ../../../python/tests/specifications/test_specifications.py
    :language: yaml
    :lines: 76-131
```

> If you would like to generate a CSV file tracking the requirement relationships in the current directory, you would execute the following command:

Reviewer: ```suggestion
If you would like to generate a CSV file tracing the requirement relationships in the current directory, you would execute the following command:
```

```bash
aac spec-csv architecture-file.aac .
```
Which would return this in the command line:
![Run Spec-Csv in Command Line](../images/spec-csv_command_line.png)

It would also generate the following three CSV files:

`Module.csv`:
![Module CSV File](../images/Module_csv.png)

`Subsystem.csv`:
![Subsystem CSV File](../images/Subsystem_csv.png)

`Other_Requirements.csv`:
![Other_Requirements CSV File](../images/Other_Requirements_csv.png)

## Definition Type References

> The two root types of definitions `spec-csv` accepts are `req` and `req_spec`

Reviewer: ```suggestion
The two root types of definitions `spec-csv` accepts are `req` and `req_spec`.
```

### req_spec Definitions

> `req_spec`s are the `specification` definitions.  `spec-csv` will generate a CSV file for each `req_spec` provided by your architecture file, with a row for each given `requirement`

Reviewer: ```suggestion
`req_spec`s are the specification definitions.  `spec-csv` will generate a CSV file for each `req_spec` provided by your architecture file, with a row for each given requirement in the `requirements` field.
```

#### req_spec Example
```yaml
req_spec:
    name: specification_name
    description: Description of Specification
    sections:
        - Section_Specification_1
        - Section_Specification_2
    parent_specs:
        - Parent_Specification
    child_specs:
        - Child Specification
    requirements:
        - "REQ-1"
        - "REQ-2"
```

In the above example, all fields except for `name` are optional fields.  The `sections`, `parent_specs`, and `child_specs` fields all take the name of another `req_spec` definition, and the `requirements` field takes the `id` of a `req` definition.

### req Definition

> `req`s are the `requirement`s of `specification`s.  In the generated CSV files, there will be a row generated for each listed `requirement`.

Reviewer: ```suggestion
`req`s are the `requirements` of `req-spec`s.  In the generated CSV files, there will be a row generated for each listed requirement.
```
